### PR TITLE
Fix status.progress not getting updated for backup

### LINKED ISSUE #6275: Progress field is missing in status of backup CR
**What steps did you take and what happened:**
Ran backup using the below command
velero backup create demo-copy-16 --include-namespaces test-csi-snapshot

**What did you expect to happen:**
The output for for "velero backup get  demo-copy-16 -o yaml"  does not contain status.progress field

**The following information will help us better understand what's going on**:

_If you are using velero v1.7.0+:_  
Please use `velero debug  --backup <backupname> --restore <restorename>` to generate the support bundle, and attach to this issue, more options please refer to `velero debug --help` 

➜  external ./velero-1.11 backup describe demo-copy-16                                                           
Name:         demo-copy-16
Namespace:    velero
Labels:       <none>
Annotations:  <none>

Phase:  New

Namespaces:
  Included:  test-csi-snapshot
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        <none>
  Cluster-scoped:  auto

Label selector:  <none>

Storage Location:  

Velero-Native Snapshot PVs:  auto

TTL:  0s

CSISnapshotTimeout:    0s
ItemOperationTimeout:  0s

Hooks:  <none>

Backup Format Version:  

Started:    <n/a>
Completed:  <n/a>

Expiration:  <nil>

Velero-Native Snapshots: <none included>

`➜  external ./velero-1.11 backup get demo-copy-19 -o yaml            
apiVersion: velero.io/v1
kind: Backup
metadata:
  annotations:
    velero.io/source-cluster-k8s-gitversion: v1.24.2
    velero.io/source-cluster-k8s-major-version: "1"
    velero.io/source-cluster-k8s-minor-version: "24"
  creationTimestamp: "2023-05-16T17:38:22Z"
  generation: 1
  labels:
    velero.io/storage-location: default
  managedFields:
  - apiVersion: velero.io/v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:velero.io/source-cluster-k8s-gitversion: {}
          f:velero.io/source-cluster-k8s-major-version: {}
          f:velero.io/source-cluster-k8s-minor-version: {}
        f:labels:
          .: {}
          f:velero.io/storage-location: {}
      f:spec:
        .: {}
        f:csiSnapshotTimeout: {}
        f:defaultVolumesToFsBackup: {}
        f:hooks: {}
        f:includedNamespaces: {}
        f:itemOperationTimeout: {}
        f:metadata: {}
        f:storageLocation: {}
        f:ttl: {}
      f:status:
        .: {}
        f:completionTimestamp: {}
        f:csiVolumeSnapshotsAttempted: {}
        f:csiVolumeSnapshotsCompleted: {}
        f:expiration: {}
        f:formatVersion: {}
        f:phase: {}
        f:startTimestamp: {}
        f:version: {}
    manager: velero-server
    operation: Update
    time: "2023-05-16T17:38:22Z"
  name: demo-copy-19
  namespace: velero
  resourceVersion: "103477"
  uid: 44bc0db7-9c4f-4796-83c7-3449449d7e93
spec:
  csiSnapshotTimeout: 10m0s
  defaultVolumesToFsBackup: false
  hooks: {}
  includedNamespaces:
  - test-csi-snapshot
  itemOperationTimeout: 1h0m0s
  metadata: {}
  storageLocation: default
  ttl: 720h0m0s
status:
  completionTimestamp: "2023-05-16T15:35:34Z"
  csiVolumeSnapshotsAttempted: 2
  csiVolumeSnapshotsCompleted: 2
  expiration: "2023-06-15T15:35:03Z"
  formatVersion: 1.1.0
  phase: Completed
  startTimestamp: "2023-05-16T15:35:03Z"
  version: 1
`

_If you are using earlier versions:_  
Please provide the output of the following commands (Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)
- `kubectl logs deployment/velero -n velero`
- `velero backup describe <backupname>` or `kubectl get backup/<backupname> -n velero -o yaml`
- `velero backup logs <backupname>`
- `velero restore describe <restorename>` or `kubectl get restore/<restorename> -n velero -o yaml`
- `velero restore logs <restorename>`

**Anything else you would like to add:**

**Environment:**

- Velero version (use `velero version`): 
Version: v1.11.0

- Velero features (use `velero client config get features`): 
- Kubernetes version (use `kubectl version`):

1.24.2

- Kubernetes installer & version:
- Cloud provider or hardware configuration:
- OS (e.g. from `/etc/os-release`):

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

### PR DESCRIPTION
**Pull request by kkothule**
Thank you for contributing to Velero!

# Please add a summary of your change
Fixed a bug where status.progress is not getting updated for backups.

# Does your change fix a particular issue?

Fixes #6275

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [x] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

File: pkg/backup/backup.go [hunk 1]
```diff
@@ -334,9 +336,13 @@ func (kb *kubernetesBackupper) BackupWithResolvers(log logrus.FieldLogger,
 			case <-ticker.C:
 				if lastUpdate != nil {
 					backupRequest.Status.Progress = &velerov1api.BackupProgress{TotalItems: lastUpdate.totalItems, ItemsBackedUp: lastUpdate.itemsBackedUp}
-					original := backupRequest.Backup.DeepCopy()
-					backupRequest.Backup.Status.Progress = &velerov1api.BackupProgress{TotalItems: lastUpdate.totalItems, ItemsBackedUp: lastUpdate.itemsBackedUp}
-					if err := kube.PatchResource(original, backupRequest.Backup, kb.kbClient); err != nil {
+					updated := backupRequest.Backup.DeepCopy()
```
Comment (blackpiglet) [inline pkg/backup/backup.go:339]: This line needs some modification. Please swap it with the previous line.

```suggestion
if lastUpdate != nil {
					updated := backupRequest.Backup.DeepCopy()
					backupRequest.Status.Progress = &velerov1api.BackupProgress{TotalItems: lastUpdate.totalItems, ItemsBackedUp: lastUpdate.itemsBackedUp}
```

**Review by blackpiglet — CHANGES_REQUESTED** [review]
Please make the modification according to the comments.

File: pkg/backup/backup.go [hunk 2]
```diff
@@ -279,10 +279,12 @@ func (kb *kubernetesBackupper) BackupWithResolvers(log logrus.FieldLogger,
...
-	backupRequest.Backup.Status.Progress.TotalItems = len(items)
-	if err := kube.PatchResource(original, backupRequest.Backup, kb.kbClient); err != nil {
+	updated := backupRequest.Backup.DeepCopy()
+	if updated.Status.Progress == nil {
+		updated.Status.Progress = &velerov1api.BackupProgress{}
+	}
+	updated.Status.Progress.TotalItems = len(items)
+	if err := kube.PatchResource(backupRequest.Backup, updated, kb.kbClient); err != nil {
 		log.WithError(errors.WithStack((err))).Warn("Got error trying to update backup's status.progress.totalItems")
 	}
```
Comment (blackpiglet) [inline pkg/backup/backup.go:289]: Another thing is we'd better update the `backupRequest.Status.Progress` after patching the backup status.

File: pkg/backup/backup.go [hunk 3]
```diff
@@ -413,9 +419,14 @@ func (kb *kubernetesBackupper) BackupWithResolvers(log logrus.FieldLogger,
 	// do a final update on progress since we may have just added some CRDs and may not have updated
 	// for the last few processed items.
 	backupRequest.Status.Progress = &velerov1api.BackupProgress{TotalItems: len(backupRequest.BackedUpItems), ItemsBackedUp: len(backupRequest.BackedUpItems)}
-	original = backupRequest.Backup.DeepCopy()
-	backupRequest.Backup.Status.Progress = &velerov1api.BackupProgress{TotalItems: len(backupRequest.BackedUpItems), ItemsBackedUp: len(backupRequest.BackedUpItems)}
-	if err := kube.PatchResource(original, backupRequest.Backup, kb.kbClient); err != nil {
+	updated = backupRequest.Backup.DeepCopy()
```
Comment (blackpiglet) [inline pkg/backup/backup.go:422]: The same line 339.

File: pkg/backup/backup.go (hunk 1 above)
Comment (kkothule) [reply pkg/backup/backup.go:338]: Done

File: pkg/backup/backup.go (hunk 2 above)
Comment (kkothule) [reply pkg/backup/backup.go:289]: Done

**Comment by kkothule** [comment]
Thank you @blackpiglet, I have update PR by addressing review comments

File: pkg/backup/backup.go (hunk 2 above)
Comment (blackpiglet) [inline pkg/backup/backup.go:289]: This line should also update the `backupRequest.Status.Progress` value.

Comment (kkothule) [reply pkg/backup/backup.go:290]: Done

**Review by blackpiglet — DISMISSED** [review]

**Comment by kkothule** [comment]
@Lyndon-Li , @shubham-pampattiwar  please review

**Comment by anshulahuja98** [comment]
@kkothule pl fix the linting 
Running golangci-lint run -c hack/../golangci.yaml
pkg/backup/backup_pv_action.go:79: unnecessary trailing newline (whitespace)

**Comment by kkothule** [comment]
> @kkothule pl fix the linting Running golangci-lint run -c hack/../golangci.yaml pkg/backup/backup_pv_action.go:79: unnecessary trailing newline (whitespace)

@anshulahuja98  The file pkg/backup/backup_pv_action.go is not part of my PR, Should i still go ahead and fix the issue?
The link work well for previous push

**Review by blackpiglet — APPROVED** [review]

**Review by shubham-pampattiwar — APPROVED** [review]

**Comment by draghuram** [comment]
In today's community call, we discussed about back-porting this PR  to 1.11 as it fixes a high impact problem. @sseago  mentioned that there is no clearly defined process for selecting fixes for back-porting, as there is for major releases.  One solution is to require two or more maintainers to agree to back-port. 

As that process is discussed, it would be nice if we can make a decision about this particular issue being part of 1.11.1. CC @blackpiglet @shubham-pampattiwar @weshayutin.

**Comment by sseago** [comment]
@draghuram Actually, we just merged that today: https://github.com/vmware-tanzu/velero/pull/6324
I didn't realize during the call that the cherry-pick PR to 1.11 was already submitted, so we were able to review and merge it today.

**Comment by blackpiglet** [comment]
@draghuram 
We are using the labels to trace PR back-porting. 
#6275 is labeled the tag `target/1.11.1`.

**Comment by draghuram** [comment]
Great. Thanks.

**Pull request merged**